### ISSUE TITLE
NullPointerException When Calling Graph API Concurrently

### ISSUE THREAD

**Issue by shahamit** [microsoftgraph/msgraph-sdk-java#241]
We are using microsoft-graph version 1.0.0 jar for making calls to the microsoft graph api. While the api calls work fine when executing in a single thread, very frequently we do get the below exception which breaks the http connection with microsoft graph servers and any further attempt to make api calls fails.

We have a pool of 5 threads that call the microsoft graph api on the "teams" resource to get channel conversations and their replies. The later is a batch request. The threads do handle "429 - Too many requests" error by waiting for "Reply-After" interval before making any further calls. During the code execution we could see that the below exception comes after we receive few such api rate limiting errors. 

Any points for further troubleshooting would be helpful. Thanks.

The exception is 

```
0420 12:52:57.817+0530 {/} ERROR [pool-3-thread-3][c.e.o.t.s.MicrosoftTeamsService-341 ] calling microsoft graph: 19:b08b3951a7654111add8d6b1c4455663@thread.skype error: {} com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:356)
	at com.microsoft.graph.http.DefaultHttpProvider.send(DefaultHttpProvider.java:197)
	at com.microsoft.graph.http.DefaultHttpProvider.send(DefaultHttpProvider.java:177)
	at com.microsoft.graph.http.BaseRequest.send(BaseRequest.java:277)
	at com.microsoft.graph.http.CustomRequest.post(CustomRequest.java:106)
	at com.example.officeintegration.teams.service.MicrosoftTeamsService.executeBatchHttpPostRequest(MicrosoftTeamsService.java:612)
Caused by: java.lang.NullPointerException: source
	at java.util.Objects.requireNonNull(Objects.java:228)
	at java.util.Scanner.<init>(Scanner.java:578)
	at com.microsoft.graph.http.DefaultHttpProvider.streamToString(DefaultHttpProvider.java:440)
	at com.microsoft.graph.http.GraphServiceException.createFromConnection(GraphServiceException.java:305)
	at com.microsoft.graph.http.DefaultHttpProvider.handleErrorResponse(DefaultHttpProvider.java:376)
	at com.microsoft.graph.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:309)
	... 208 common frames omitted
```

**Comment by NakulSabharwal** [comment]
Hi @shahamit , thanks for creating this issue. Can you tell what URL you are passing in CustomRequest along with Graph Model you what it to typecast to ? Also, if possible can you paste code snippet of request you are calling.

**Comment by shahamit** [comment]
Hi @NakulSabharwal 
We are calling graph api's on the "teams" resource. We fetch latest 20 conversations from all channels and for each conversation we fetch its replies. The latter is done through a batch request where we batch 20 api calls for fetching replies of 20 conversations. 

Below is the pseudo code snippet that schedules a thread to execute a recurring task periodically. The task internally uses a pool of 5 threads to make the api calls for multiple channels in multiple teams. 
I have tried to simplify the code as much as possible. Please let me know if you have more questions

```
        int noOfThreads = 5;
        ExecutorService ioBound = Executors.newFixedThreadPool(noOfThreads);

        Runnable runnabledelayedTask = () -> executeAutoRefresh(teamIds, accessToken, noOfThreads, ioBound);
        ScheduledExecutorService scheduledPool = Executors.newScheduledThreadPool(1);

        scheduledPool.scheduleWithFixedDelay(runnabledelayedTask, 1, 10, TimeUnit.SECONDS);

    private void executeAutoRefresh(List<String> teamIds, String accessToken, int noOfThreads, ExecutorService ioBound) {
        List<CompletableFuture> futures = newArrayList();

        for (String teamId : teamIds) {
            for (String channelId : conversationCacheService.getTeamChannelIds(teamId)) {
                futures.add(CompletableFuture.runAsync(() -> refreshOneChannel(accessToken, teamId, channelId), ioBound));
            }
        }


        CompletableFuture<Void> last = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).thenRun(() -> {
            LOGGER.info("Done")
	});

        last.join();
    }


    public void refreshOneChannel(String accessToken, String teamId, String channelId) {
        try {
            loadChannelConversations(teamId, channelId, accessToken);            
        } catch (Exception e) {
            LOGGER.error("error", e);
        }
    }

    public void loadChannelConversations(String teamId, String channelId, String accessToken) throws IOException {
        
        try {
            JsonObject conversations = getBetaGraphServiceClient(accessToken).customRequest(format("/teams/%s/channels/%s/messages", getURLEncodedString(teamId), getURLEncodedString(channelId))).buildRequest().get();
            updatedConversationDTO = getConversationDTO(teamId, channelId, conversations);
            if (!updatedConversationDTO.getChannelMessages().isEmpty()) {
                List<String> threadIds = getMessageThreadIds(updatedConversationDTO);
                loadChannelConversationsReplies(teamId, channelId, threadIds, accessToken);
            }
        } catch (GraphServiceException e) {
            if (e.getResponseCode() == TOO_MANY_REQUESTS.value()) {
                LOGGER.error("TooManyRequests error from Conversations API.");
                long retryAfterSeconds = parseLong(getRetryAfterValueFromResponse(e.getMessage(true)));
                waitForGraphThrottleLimit(retryAfterSeconds);
                loadChannelConversations(teamId, channelId, accessToken); //call again
            }
        }        
    }


    private void loadChannelConversationsReplies(String teamId, String channelId, List<String> messageThreadIds, String accessToken) throws IOException {
        JsonObject batchJSONObject = createConversationRepliesBatchHttpPostRequest(teamId, channelId, messageThreadIds, accessToken);
        //check for error codes and retry after "reply-after" interval like done above. Removed code for simplicity purpose
    }


    private JsonObject createConversationRepliesBatchHttpPostRequest(String teamId, String channelId, List<String> messageThreadIds, String accessToken) {
        List<HttpRequest> requests = new ArrayList<>();
        int id = 1;
        for (String messageThreadId : messageThreadIds) {
            requests.add(new HttpRequest(valueOf(id), GET, format("/teams/%s/channels/%s/messages/%s/replies", getURLEncodedString(teamId), getURLEncodedString(channelId), messageThreadId)));
            id++;
        }
        JsonObject jsonObject = new JsonObject();
        jsonObject.add(REQUESTS, new Gson().toJsonTree(requests));
        JsonObject replies = executeBatchHttpPostRequest(accessToken, jsonObject);
        return replies;
    }

    public JsonObject executeBatchHttpPostRequest(String accessToken, JsonObject jsonObject) {
        return getBetaGraphServiceClient(accessToken).customRequest("/$batch")
                .buildRequest().post(jsonObject);
    }
```

**Comment by shahamit** [comment]
@NakulSabharwal - did you get to the root cause and the fix for this issue? It is kind of a blocker to us. 
Feel free to let me know if you need any help in reproducing the issue.

**Comment by NakulSabharwal** [comment]
@shahamit Which endpoint are you calling concurrently ? Can you elaborate on that

**Comment by shahamit** [comment]
@NakulSabharwal - I have already highlighted in the above code snippet. It is the batch endpoint. I doubt though that the error is specific to any endpoint. Which endpoints did you try it with?

**Comment by NakulSabharwal** [comment]
I am not able to repro this issue by calling the batch endpoint concurrently using - 
```
for(int i=0;i<1000;i++) {
			client.customRequest("/$batch").buildRequest().post(o, new ICallback<JsonObject>() {
				
				@Override
				public void success(JsonObject result) {
					//System.out.println(result);
				}
				
				@Override
				public void failure(ClientException ex) {
					System.out.println(ex);
				}
			});
		}
```
but as I can see in the error stacktrace provided above it seems the inputstream is null and hence throws an error.

**Comment by shahamit** [comment]
@NakulSabharwal - As I already stated in the issue subject and description, this issue reproduces when you execute the api calls concurrently and this is a reason why I also pasted the sample code in a later comment. 
The above code that you tried doesn't have any concurrency. I hope you have understood the issue that I have reported here. Please let me know if you need more details.

**Comment by NakulSabharwal** [comment]
@shahamit The code I pasted is making multiple requests to batch endpoint Asyncronously. Can you put a breakpoint at handleErrorResponse(request, serializable, connection); in DefaultHttpProvider and tell the status code you are receiving. It would be very helpful.

**Comment by shahamit** [comment]
@NakulSabharwal - Where can I find the status code property? Which object would have it? Obviously request doesn't have it. 
Also it seems this method is called in successful flows too since the code execution stuck at this breakpoint even when there were no errors. Could you please detail on the code flow?

**Comment by shahamit** [comment]
@NakulSabharwal - This issue is very critical to us. Can we connect over a VC meeting to speed up the root cause identification?

**Comment by NakulSabharwal** [comment]
@shahamit The error was due to empty error response body. I have added null check for that scenario. Can you clone the repo and test if it is working fine in your use case ? If not then we have a VC.

**Comment by shahamit** [comment]
@NakulSabharwal - Sure I will give it a try. From the code fix, it seems microsoft graph returns an empty response in some cases. Do you know when could this happen?

**Comment by NakulSabharwal** [comment]
@shahamit - Did you give it a try ? I am not sure about empty error response, will let you know.

**Comment by shahamit** [comment]
@NakulSabharwal - Yes the first couple of runs didn't give the error. This looks like non-harmful change. Can you merge it with master and share a master build?
Thanks!

**Comment by NakulSabharwal** [comment]
@shahamit This issue has been fixed and release version is 1.4.0. Please reopen if you still facing any problem.
